File: graylog2-web-interface/src/views/logic/TimeRange.test.ts
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import type { TimeRangeQueryParameter } from 'views/logic/TimeRange';
import { timeRangeToQueryParameter, timeRangeFromQueryParameter } from 'views/logic/TimeRange';
import type { RelativeTimeRange, AbsoluteTimeRange, KeywordTimeRange } from 'views/logic/queries/Query';

describe('TimeRange', () => {
  describe('timeRangeToQueryParameter', () => {
    it('handles relative time ranges', () => {
      expect(timeRangeToQueryParameter({ type: 'relative', range: 300 }))
        .toEqual({ rangetype: 'relative', relative: '300' });

      expect(timeRangeToQueryParameter({ type: 'relative', from: 300 }))
        .toEqual({ rangetype: 'relative', from: '300' });

      expect(timeRangeToQueryParameter({ type: 'relative', from: 300, to: 150 }))
        .toEqual({ rangetype: 'relative', from: '300', to: '150' });

      expect(() => timeRangeToQueryParameter({ type: 'relative' } as RelativeTimeRange)).toThrow(/Unexpected time range/);
    });

    it('handles absolute time ranges', () => {
      expect(timeRangeToQueryParameter({ type: 'absolute', from: '2022-08-03T11:52:00+01:00', to: '2022-08-04T11:52:00+01:00' }))
        .toEqual({ rangetype: 'absolute', from: '2022-08-03T11:52:00+01:00', to: '2022-08-04T11:52:00+01:00' });

      expect(timeRangeToQueryParameter({ type: 'absolute' } as AbsoluteTimeRange))
        .toEqual({ rangetype: 'absolute', from: undefined, to: undefined });
    });

    it('handles keyword time ranges', () => {
      expect(timeRangeToQueryParameter({ type: 'keyword', keyword: 'yesterday' }))
        .toEqual({ rangetype: 'keyword', keyword: 'yesterday' });

      expect(timeRangeToQueryParameter({ type: 'keyword' } as KeywordTimeRange))
        .toEqual({ rangetype: 'keyword', keyword: undefined });
    });
  });

  describe('timeRangeFromQueryParameter', () => {
    it('handles relative time ranges', () => {
      expect(timeRangeFromQueryParameter({ rangetype: 'relative', relative: '300' }))
        .toEqual({ type: 'relative', range: 300 });

      expect(timeRangeFromQueryParameter({ rangetype: 'relative', from: '300' }))
        .toEqual({ type: 'relative', from: 300 });

      expect(timeRangeFromQueryParameter({ rangetype: 'relative', from: '300', to: '150' }))
        .toEqual({ type: 'relative', from: 300, to: 150 });

      expect(() => timeRangeFromQueryParameter({ rangetype: 'relative' } as TimeRangeQueryParameter)).toThrow(/Invalid relative range specified/);
    });

    it('handles absolute time ranges', () => {
      expect(timeRangeFromQueryParameter({ rangetype: 'absolute', from: '2022-08-03T11:52:00+01:00', to: '2022-08-04T11:52:00+01:00' }))
        .toEqual({ type: 'absolute', from: '2022-08-03T11:52:00+01:00', to: '2022-08-04T11:52:00+01:00' });

      expect(timeRangeFromQueryParameter({ rangetype: 'absolute', from: undefined, to: undefined }))
        .toEqual({ type: 'absolute' } as AbsoluteTimeRange);
    });

    it('handles keyword time ranges', () => {
      expect(timeRangeFromQueryParameter({ rangetype: 'keyword', keyword: 'yesterday' }))
        .toEqual({ type: 'keyword', keyword: 'yesterday' });

      expect(timeRangeFromQueryParameter({ rangetype: 'keyword', keyword: undefined }))
        .toEqual({ type: 'keyword' });
    });
  });
});

File: graylog2-server/src/test/java/org/graylog2/plugin/indexer/searches/timeranges/TimeRangeTest.java
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package org.graylog2.plugin.indexer.searches.timeranges;

import org.graylog.plugins.views.search.timeranges.OffsetRange;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.junit.jupiter.api.Test;

import static org.graylog.testing.jackson.JacksonSubtypesAssertions.assertThatDto;

class TimeRangeTest {
    @Test
    void subtypes() {
        final var now = DateTime.now(DateTimeZone.UTC);

        final var absoluteRange = AbsoluteRange.create(now, now);
        final var relativeRange = RelativeRange.create(500);
        final var keywordRange = KeywordRange.create("yesterday", "UTC");
        final var offsetRange = OffsetRange.Builder.builder()
                .offset(1)
                .source("foo")
                .build();

        assertThatDto(absoluteRange)
                .doesNotSerializeWithDuplicateFields()
                .deserializesWhenGivenSupertype(TimeRange.class);
        assertThatDto(relativeRange)
                .doesNotSerializeWithDuplicateFields()
                .deserializesWhenGivenSupertype(TimeRange.class);
        assertThatDto(keywordRange)
                .doesNotSerializeWithDuplicateFields()
                .deserializesWhenGivenSupertype(TimeRange.class);
        assertThatDto(offsetRange)
                .doesNotSerializeWithDuplicateFields()
                .deserializesWhenGivenSupertype(TimeRange.class);
    }
}
